src/libappimage/desktop_integration/Thumbnailer.h

```diff
@@ -0,0 +1,63 @@
+// system
+#include <string>
+
+// local
+#include "integrator/DesktopIntegrationResources.h"
+
+namespace appimage {
+    namespace desktop_integration {
+        /**
+         * Thumbnails generator for AppImage files
+         *
+         * Follows the Thumbnail Managing Standard by FreeDesktop
+         * https://specifications.freedesktop.org/thumbnail-spec/0.8.0/index.html
+         */
+        class Thumbnailer {
+            std::string xdgCacheHome;
+
+        public:
+            explicit Thumbnailer(const std::string& xdgCacheHome = "");
+
+            /**
+             * @brief Generate thumbnails for the given <appImagePath>
+             *
+             * Thumbnail generation is performed according to the Freedesktop specification.
+             * Two images of 128x128 and 256x256 will be placed at "$XDG_CACHE_HOME/thumbnails/normal" and
+             * "$XDG_CACHE_HOME/thumbnails/large/" respectively. The thumbnails name will be formed by a md5 sum of
+             * the absolute canonical URI for the original file whit ".png" as extension.
+             *
+             * Full FreeDesktop Thumbnails spec: https://specifications.freedesktop.org/thumbnail-spec/0.8.0/x227.html
+             *
+             * @param appImagePath
+             */
+            void create(const std::string& appImagePath);
+
+            virtual ~Thumbnailer();
+
+            void remove(const std::string& appImagePath);
+
+        private:
+            boost::filesystem::path getNormalThumbnailPath(const std::string& canonicalPathMd5) const;
```

Comment:
This is not a public header, therefore, there is no need for PImpl here. It will only make the implementation more complex and add runtime overhead.

Reply:
If you use a pattern in a code base, do it everywhere. That's just consistency. There's no real overhead, it's just cleaner and easier to develop since everything's in the source file, just like everywhere else.

Reply:
Software design patterns should not be used without a real reason or just because they are coll or used elsewhere. The rationale behind this pattern is to provide a consistent API and ABI to the class users. This is a private (in what respect to the project) class, therefore, there is no need to use this pattern. 

There is a real overhead on using it. An extra resolution on every member access will be required and there will be another class to deal with. Let's be careful about not overengineering this. I've fallen a lot of times in this issue and I would like to stay away from it.